prescription/store/mongo/prescription_session.go

```diff
@@ -18,38 +22,195 @@ type PrescriptionSession struct {
+	if ctx == nil {
+		return nil, errors.New("context is missing")
+	}
+	if p.IsClosed() {
+		return nil, errors.New("session closed")
+	}
+	if accessCode == "" {
+		return nil, errors.New("access code is missing")
+	}
+
+	now := time.Now()
+	logger := log.LoggerFromContext(ctx).WithFields(log.Fields{"accessCode": accessCode})
+
+	selector := bson.M{
+		"accessCode": accessCode,
+		"patientId":  nil,
+		"state":      prescription.StateSubmitted,
+	}
+
+	prescr := &prescription.Prescription{}
+	err := p.C().Find(selector).One(prescr)
+	logger.WithFields(log.Fields{"duration": time.Since(now) / time.Microsecond}).WithError(err).Debug("GetUnclaimedPrescription")
+	if err == mgo.ErrNotFound {
+		return nil, nil
+	} else if err != nil {
+		return nil, errors.Wrap(err, "unable to find unclaimed prescription")
+	}
+
+	return prescr, nil
+}
+
+func (p *PrescriptionSession) DeletePrescription(ctx context.Context, clinicianID string, id string) (bool, error) {
+	if ctx == nil {
+		return false, errors.New("context is missing")
+	}
+	if p.IsClosed() {
+		return false, errors.New("session closed")
+	}
+	if clinicianID == "" {
+		return false, errors.New("clinician id is missing")
+	}
+	if id == "" {
+		return false, errors.New("prescription id is missing")
+	} else if !bson.IsObjectIdHex(id) {
+		return false, nil
+	}
+
+	now := time.Now()
+	logger := log.LoggerFromContext(ctx).WithFields(log.Fields{"clinicianId": clinicianID, "id": id})
+
+	selector := bson.M{
+		"_id": bson.ObjectIdHex(id),
+		"$or": []bson.M{
+			{"prescriberId": clinicianID},
+			{"createdUserId": clinicianID},
+		},
+		"state": bson.M{
+			"$in": []string{prescription.StateDraft, prescription.StatePending},
+		},
+		"deletedTime": nil,
+	}
+	update := bson.M{
+		"$set": bson.M{
+			"deletedUserId": clinicianID,
+			"deletedTime":   now,
+		},
+	}
+
+	err := p.C().Update(selector, update)
+	logger.WithFields(log.Fields{"duration": time.Since(now) / time.Microsecond}).WithError(err).Debug("DeletePrescription")
+	if err == mgo.ErrNotFound {
+		return false, nil
+	} else if err != nil {
+		return false, errors.Wrap(err, "unable to delete prescription")
+	} else {
+		return true, nil
+	}
+}
+
+func newMongoSelectorFromFilter(filter *prescription.Filter) bson.M {
+	selector := bson.M{}
+	if filter.GetClinicianID() != "" {
+		selector["$or"] = []bson.M{
+			{"prescriberId": filter.GetClinicianID()},
+			{"createdUserId": filter.GetClinicianID()},
+		}
+	}
+	if filter.PatientID != "" {
+		selector["patientId"] = filter.PatientID
+	}
+	if filter.PatientEmail != "" {
+		selector["latestRevision.attributes.email"] = filter.PatientEmail
+	}
+	if filter.ID != "" {
+		if bson.IsObjectIdHex(filter.ID) {
+			selector["_id"] = bson.ObjectIdHex(filter.ID)
+		} else {
+			selector["_id"] = nil
+		}
+	}
+	if filter.State != "" {
+		selector["state"] = filter.State
+	}
+	if filter.CreatedTimeStart != nil {
+		selector["createdTime"] = bson.M{"$gt": filter.CreatedTimeStart}
```

Comment:
This interval is open on both ends.  Is that intentional. Do you want to use a half-open interval?

Reply:
Not intentional - I'll get that fixed.